### PR TITLE
feat: Flux package manager that can renovate HelmRelease manifests

### LINKED ISSUE #5984: Support Renovating HelmRelease CRD
**What would you like Renovate to be able to do?**
Raise Pull Requests for updates to HelmRelease CRDs with latest helm charts versions. 

**Describe the solution you'd like**
Extend the solution of Helm requirements.yaml to renovate HelmRelease [CRD](https://github.com/fluxcd/helm-operator/blob/master/deploy/crds.yaml). For example, actual version that needs to be updated is [here](https://github.com/hmcts/cnp-flux-config/blob/master/k8s/admin/traefik/traefik.yaml#L13) 

**Describe alternatives you've considered**
No other alternatives available at the moment.

**Additional context**
Reuse as described in https://github.com/renovatebot/renovate/issues/3414#issuecomment-541106635

### PR DESCRIPTION
**Pull request by danports**
Adds a new `flux` package manager that parses `HelmRelease` and `HelmRepository` manifests. Mostly code adapted from the `helmfile`, `gitlabci`, and `kubernetes` managers. Closes #5984.

- [x] Update docs
- [x] Add unit tests
- [x] Test on a real repository
- [x] Add more tests to achieve 100% code coverage
- [x] Review `SkipReason` returned when failing to match a  `HelmRelease` to a `HelmRepository`

This was my first time writing TypeScript, and it was really quite pleasant! 🙂

### CODE REVIEW

**Comment by anthr76** [comment]
Closes: https://github.com/renovatebot/renovate/issues/12139

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:25]: Could be good to check `is.nonEmptyArray(resources)` before this, because it may not even be an array. And if it's an empty array, we may as well return null too.

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:28]:
```suggestion
        if (resource.apiVersion?.startsWith('helm.toolkit.fluxcd.io/')) {
```

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:26]:
```suggestion
    switch (resource?.kind) {
```

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:33]:
```suggestion
        if (resource.apiVersion?.startsWith('source.toolkit.fluxcd.io/')) {
```

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:50]: These need some defensive checks (`?.`) unless you have pre-validated them to exist

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs some defensive non-null checks added in quite a few places, unless the response is first fully validated to conform to the TS type(s)

File: lib/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/manager/flux/readme.md:3]:
```suggestion
The `flux` manager will only extract `HelmRelease` resources linked to `HelmRepository` sources.
`HelmRelease` resources linked to other kinds of sources like `GitRepository` or `Bucket` will be ignored.
```

We can drop the `Note that` part, and use one sentence per line.

File: lib/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/manager/flux/readme.md:5]:
```suggestion
For the `flux` manager to properly link `HelmRelease` and `HelmRepository` resources, _both_ must have their `namespace` explicitly set in their `metadata`.
Dependencies will not be extracted from resources where `metadata.namespace` is omitted.
However, the namespace can be omitted from the `HelmRelease` resource's `spec.chart.spec.sourceRef` property.
```

Simplify start of sentence, emphasize the _both_ part, use one sentence per line.

File: lib/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/manager/flux/readme.md:7]:
```suggestion
The `flux` manager has no `fileMatch` default patterns, so it won't match any files until you configure it with a pattern.
This is because there is no commonly accepted file/directory naming convention for Flux manifests and we don't want to check every single `*.yaml` file in repositories just in case any of them contain Flux definitions.
```

One sentence per line.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:50]: Interesting. I had assumed that TS would validate objects during a type cast, but that was a misunderstanding - and come to think of it, that would be a lot of code for the compiler to generate. I'll fix this.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:25]: I think we're fine here, since according to the `js-yaml` docs, `loadAll` always returns an array (with one element in the common case of a doc with a single resource), and empty arrays are handled later in the flow.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:28]: Added with a test.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:33]: Added with a test.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:26]: Added with a test.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:58]: Added these checks along with a slew of additional tests. This exercise reminded me of both the good and the bad of the years I spent practicing TDD... 🙂

File: lib/manager/flux/extract.spec.ts
Comment (viceice) [inline lib/manager/flux/extract.spec.ts:23]: Please use `toEqual` matcher, as the result.is a pretty small object

File: lib/manager/flux/extract.spec.ts
Comment (viceice) [inline lib/manager/flux/extract.spec.ts:19]:
```suggestion
```
Not required, as the `set` will do it also.

File: lib/manager/flux/extract.spec.ts
Comment (viceice) [inline lib/manager/flux/extract.spec.ts:177]: See other comment.

File: lib/manager/flux/extract.ts
Comment (viceice) [inline lib/manager/flux/extract.ts:14]:
```suggestion
function readManifest(content: string): FluxManifest | null {
```

Please add `lib/flux` to `tsconfig.strict.json`
If there are transitive errors, only try to fix issues in flux and remove from strict afterwards. 🤗

File: lib/manager/flux/extract.ts
Comment (viceice) [inline lib/manager/flux/extract.ts:29]: Move the API version check before the switch to make code less verbose

File: lib/manager/flux/index.ts
Comment (viceice) [inline lib/manager/flux/index.ts:4]:
```suggestion
```

File: lib/manager/flux/index.ts
Comment (viceice) [inline lib/manager/flux/index.ts:1]:
```suggestion
export { extractAllPackageFiles, extractPackageFile } from './extract';
```

File: lib/manager/flux/extract.spec.ts
Comment (viceice) [inline lib/manager/flux/extract.spec.ts:6]:
```suggestion
import { extractAllPackageFiles, extractPackageFile } from '.';
```
Coverage. 😉

**Comment by anthr76** [comment]
Any chance we can cover a flux kustomization in this PR? It would resolve https://github.com/renovatebot/renovate/issues/12139

File: lib/manager/flux/extract.spec.ts
Comment (danports) [reply lib/manager/flux/extract.spec.ts:19]: OK, I had copied this from some of the `gitlabci` manager tests. I'll remove it.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:29]: Can you clarify? The arguments to `startsWith` are different since the resources are in different API groups so I'm not sure how to make this code less verbose. 🤔

**Comment by danports** [comment]
@anthr76 I read through #12139 but it's not clear to me what needs to be renovated in `Kustomization` resources. Can you clarify in that issue? I've only used `Kustomization` manifests at a very basic level in Flux, so perhaps I just haven't run into that use case yet. My preference would be to land this PR and then consider extending the Flux manager in future PRs.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:14]: I added `lib/manager/flux/**/*.ts` to `tsconfig.strict.json` and ran `yarn build` and it succeeded - I'm guessing that means we're good...if we're not I'm sure the build will yell at me shortly. 🙂

Comment (danports) [reply lib/manager/flux/extract.ts]: Oh, I had to run `yarn null-check`...the output from the broken build only shows a few warnings in `flux/extract.spec.ts`, so I think we're good here.

**Comment by anthr76** [comment]
Sure

Here's a few usecase examples.. Looking back I suppose this is more or less for a gitReository spec

1. https://github.com/anthr76/infra/blob/main/gitops/base/monitoring/kube-prometheus-stack/crds/kustomize.yaml#L9-L17
2. https://github.com/anthr76/infra/blob/ab4463314fe1cca8d5af2b440a30e2281f7ab7ba/gitops/base/database/kubegres/git-repository.yaml#L9-L13
3. https://github.com/anthr76/infra/blob/ab4463314fe1cca8d5af2b440a30e2281f7ab7ba/gitops/base/kube-system/kubelet-serving-cert-approver/kustomize.yaml#L9-L12

I glue these together today with a regex manager https://github.com/anthr76/infra/blob/ab4463314fe1cca8d5af2b440a30e2281f7ab7ba/.github/renovate.json5#L38-L49 basically looking at it's source and updating the git tags most of the time.

**Comment by danports** [comment]
Got it. I don't currently use `GitRepository` resources in my Flux repositories so that wouldn't be a priority for me but is certainly something that could be added in a future PR.

**Comment by rarkins** [comment]
@anthr76 can you create a feature request for the additional features? Either create once this is merged or assume this PR will be merged soon!

File: lib/manager/flux/extract.ts
Comment (rarkins) [reply lib/manager/flux/extract.ts:29]: @viceice were you thinking for DRY reasons? Because the two strings are difference, so I'm also confused

File: lib/manager/flux/extract.ts
Comment (viceice) [inline lib/manager/flux/extract.ts:121]:
```suggestion
```
Will be logged later anyways

File: lib/manager/flux/extract.ts
Comment (viceice) [inline lib/manager/flux/extract.ts:99]:
```suggestion
      logger.debug({ file }, 'Empty Flux manifest');
```
File exists at that stage.

File: lib/manager/flux/extract.ts
Comment (viceice) [inline lib/manager/flux/extract.ts:26]: Add default case with trace log?

File: lib/manager/flux/extract.ts
Comment (viceice) [reply lib/manager/flux/extract.ts]: If no errors occured, please add and commit again.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:99]: Removed this whole `if` block - there's no real need to log empty files or treat them as a special case.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:26]: I don't think it's necessary to log the kinds of all other resources the parser might run into - there are plenty of other types in the Flux toolkit that we're ignoring.

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:88]: `extractAllPackageFiles` is only necessary if package files need to be correlated amongst each other. Is that the case here? It seems `extractPackageFile` is also defined but not referenced internally

**Review by rarkins — COMMENTED** [review]
Confusion over extractAllPackageFiles

File: lib/manager/flux/extract.ts
Comment (viceice) [reply lib/manager/flux/extract.ts:88]: `extractPackageFile` is required for auto-replace 😕

Comment (rarkins) [reply lib/manager/flux/extract.ts:88]: I'm wondering if it's the only one required?

Comment (danports) [reply lib/manager/flux/extract.ts:88]: Yes, we need `extractAllPackageFiles` since we have to read all `HelmRelease` and `HelmRepository` manifests to correlate `HelmReleases` resources to `HelmRepository` resources. And yes, @viceice is correct, we need `extractPackageFile` for auto-replace - based on my testing, it is the only function required for auto-replace to work (it's just used to validate that the version replacement was successful).

File: lib/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/manager/flux/readme.md:4]:
```suggestion
`HelmRelease` resources linked to other sources like `GitRepository` or `Bucket` will be ignored.
```

File: lib/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/manager/flux/readme.md:9]:
```suggestion
1. The `HelmRelease` resource must either have its `metadata.namespace` property set or its `spec.chart.spec.sourceRef.namespace` property set
2. The referenced `HelmRepository` resource must have its `metadata.namespace` property set
```

We don't use punctuation at the end of a list.

File: lib/manager/flux/readme.md
Comment (HonkingGoose) [inline lib/manager/flux/readme.md:26]:
```suggestion
If instead you have all your Flux manifests inside a `flux/` directory, you would add this:
```

Explaining what `them` refers makes the sentence easier to understand, when you read the sentence on its own.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

**Review by viceice — COMMENTED** [review]
LGTM, waiting for doc fixes 🤗

File: lib/manager/flux/readme.md
Comment (danports) [reply lib/manager/flux/readme.md:4]: `GitRepository` (and `Bucket`) is a `kind` of Flux source - it's not a source in itself, which is why I phrased it this way.

**Review by rarkins — APPROVED** [review]

**Review by HonkingGoose — APPROVED** [review]
I'm happy with the docs now. 😄

**Review by viceice — APPROVED** [review]

**Comment by ChipWolf** [comment]
🚀

**Comment by danports** [comment]
How long does it take new builds to make it to Docker Hub? Was hoping to update my GitLab Renovate runner to use this release, but I don't see it on Docker Hub yet.

**Comment by rarkins** [comment]
Usually a few hours. Keep an eye on the renovatebot/docker-renovate-full repo

**Pull request merged**
